p1_runner/data_source.py

```diff
@@ -137,26 +139,35 @@ def __init__(
             # The socket buffer can be made large enough that we don't need a reader thread.
             self.socket_in.setsockopt(socket.SOL_SOCKET, socket.SO_SNDBUF, MAX_DATA_BUFFER_SIZE)
             self.socket_in.setblocking(False)
+            enable_socket_timestamping(self.socket_in, enable_sw_timestamp=True, enable_hw_timestamp=False) # type: ignore
 
         self.rx_log = rx_log
 
     def write(self, data: bytes):
         self.socket_out.sendall(data)
 
-    def read(self, size: int, timeout=RESPONSE_TIMEOUT) -> bytes:
-        data = b''
+    def read(self, size: int, timeout=RESPONSE_TIMEOUT, return_any: bool = False) -> bytes:
+        all_data = b''
         start_time = time.time()
         remaining = timeout
-        while len(data) < size and remaining >= 0:
+        while len(all_data) < size and remaining >= 0:
             ready = select.select([self.socket_in], [], [], remaining)
             if ready[0]:
-                data += self.socket_in.recv(size - len(data))
+                data, ancdata, _, _ = self.socket_in.recvmsg(size - len(data), 1024)
+                kernel_ts, _, _ = parse_timestamps_from_ancdata(ancdata)
+                if self.rx_log:
+                    if kernel_ts and isinstance(self.rx_log, LogManager):
+                        self.rx_log.write_with_timestamp(data, kernel_ts)
```

Comment:
Should extracting and logging timestamps be optional? I suspect the extracting doesn't add _that_ much overhead, but writing the timestamp file out might be unexpected if you didn't ask for it?

Reply:
That is already controlled by the LogManager. If you don't enable timestamp logging explicitly, it won't be logged.